### DISCUSSION TITLE
Mirror client TLS configuration in server ClientHello?

### DISCUSSION THREAD

**Discussion by daMatz** [mitmproxy/mitmproxy#5748] [Q&A]
Is there a way to replicate/mirror the TLS configuration of the original package with mitmproxy?
I know, that we can set e.g. the ciphers via the `ciphers_server` option, but there are limitations like https://github.com/mitmproxy/mitmproxy/issues/4260.

If in the current state mitmproxy is not able to perform this, would it be feasible to manually manipulate the TLS config within the `ClientHello` package and recalculate certain length checks / checksums or is there a easier way to achieve this?

Happy to discuss possible solutions here :)

**Comment by mhils** [comment]
See https://github.com/mitmproxy/mitmproxy/issues/4575 and https://github.com/mitmproxy/mitmproxy/pull/5157.

In short, OpenSSL currently does not provide us with the level of control we'd need to exactly mirror a ClientHello. It'd be great to have, but not easy to implement as it seems.

**Reply by mhils** [reply]
AFAIK, OpenSSL does not have an API to set/use arbitrary ClientHellos.

**Reply by mhils** [reply]
AFAIK the current OpenSSL APIs don't allow you to for instance specify GREASE ciphersuites (see the discussion in https://github.com/mitmproxy/mitmproxy/pull/5157), so mirroring those from a client is not possible at the moment (I think).

**Comment by fedosgad** [comment]
Hi! I'm the author of #5157.
You may be interested in https://github.com/fedosgad/mirror_proxy (I've written it for exactly the same purpose).

**Reply by mhils** [reply]
Very nice work, thanks for sharing! 🍰
